src/Build/Instance/ProjectInstance.cs

```diff
@@ -2710,5 +2771,30 @@ private void CreatePropertiesSnapshot(Evaluation.Project.Data data, bool isImmut
                 _properties.Set(instance);
             }
         }
+
+        /// <summary>
+        /// Tracks this property read if it's from an environment variable or uninitialized.
+        /// </summary>
+        /// <param name="name">The name of the property being read.</param>
+        private void TrackPropertyRead(string name, bool propertyMissing)
+        {
+            if (string.IsNullOrEmpty(name))
+                return;
```

Comment:
does a null or empty name has any special semantics?
Should this be an assert instead?

Reply:
I saw one instance of this and that's why I added it. MSBuild looks up a property called "InnerBuildProperty". If that isn't present, an empty string is returned and it then attempts to look up the value for that property (which is an empty string). This appeared to be current behavior and so instead of an assert I just checked explicitly.

Reply:
This shows an interesting consequence of the current design: it tracks properties reads during evaluation, and also after evaluation, when the ProjectInstance is interogated for values via its API. Would it make sense to restrict this to recording only the reads that happen during evaluation? For me it makes sense to restrict it, otherwise it means that you may get different property read data every time you ask for it.